Fix force update translations 05 cross site request forgery

rocket-martue opened the pull request:
Fix CSRF vulnerability and improve PHP 8.2 compatibility

rocket-martue committed:
Fix CSRF vulnerability in plugin and theme translation updates

- Add nonce verification to plugin translation update
- Add nonce verification to theme translation update
- Add permission checks (update_plugins, edit_theme_options)
- Change theme update from GET to POST method
- Add user action requirement (button click)
- Improve code documentation and visibility

This fixes the CSRF vulnerability reported in Wordfence:
https://www.wordfence.com/threat-intel/vulnerabilities/wordpress-plugins/force-update-translations/force-update-translations-05-cross-site-request-forgery

rocket-martue committed:
Fix PHP 8.2 compatibility and improve code quality

- Add root_slug property declaration to GP_Locale class (fixes PHP 8.2 deprecated warning)
- Change admin notices from inline to is-dismissible for better visibility
- Fix WPCS (WordPress Coding Standards) violations in locales.php:
  - Fix indentation and spacing
  - Use self:: instead of GP_Locales:: for static calls
  - Add parentheses to class instantiation
  - Fix ternary operator spacing
  - Fix foreach spacing

rocket-martue committed:
Address Copilot AI review: fix admin_notices array structure and strengthen path traversal protection

rocket-martue commented on inc/plugins.php:
```diff
@@ -40,38 +44,59 @@ function plugin_action_links( $actions, $plugin_file ) {
+			return;
+		}
+
+		if ( ! preg_match( '/^([a-zA-Z0-9-_]+)\/([a-zA-Z0-9-_.]+.php)$/', $plugin_file, $plugin_slug ) ) {
```
Thanks for the review! I've addressed both issues:

1. **Admin notices array structure**: Unified all error keys to use 'error' instead of specific keys like 'csrf_error', 'permission_error', etc. to match the expected pattern in admin_notices() iterator.

2. **Path traversal protection**: 
   - Strengthened regex pattern to remove dots from filename matching
   - Added realpath() validation to ensure the plugin file exists and is within WP_PLUGIN_DIR
   - Added strpos() check to prevent directory traversal attacks

Changes pushed in commit 48993af.

rocket-martue committed:
Replace check_admin_referer with wp_verify_nonce for better error handling in themes.php

rocket-martue commented:
Thanks for the additional review! I've addressed all three issues:

1. **themes.php - Nonce verification**: Replaced `check_admin_referer()` with `wp_verify_nonce()` to provide a custom, user-friendly error message instead of the generic wp_die() screen. Also added proper sanitization with `sanitize_text_field( wp_unslash() )`. (Commit 6d2d718)

2. **plugins.php - Admin notices array structure**: Unified all error keys to use `'error'` instead of specific keys like `'csrf_error'`, `'permission_error'`, etc. to match the expected pattern in the `admin_notices()` iterator. (Commit 48993af)

3. **plugins.php - Path traversal protection**: Strengthened the regex pattern to prevent directory traversal by removing dots from the filename pattern and added `realpath()` validation with `strpos()` check to ensure the plugin file exists and is within `WP_PLUGIN_DIR`. (Commit 48993af)

All changes have been tested and pushed. The code now follows consistent patterns across both plugin and theme update handlers.

mayukojpn commented:
Thank you @rocket-martue ! I finally started reviewing the pull request 😭 I'm so sorry it took me this long to get to it, especially after you took the time to send this patch! 🙏

I was reviewing the changes and just realized something important: the `locales.php` file is actually a complete copy from the GlotPress plugin! So I'm thinking it might be better to use the latest version from GlotPress directly:
https://github.com/GlotPress/GlotPress/blob/develop/locales/locales.php

What do you think? I apologize for bringing this up after you've already done so much work on it 😭

I'm also wondering if there's a better practice for including PHP files from other plugins - this has been on my mind for a while.

Thanks again for your patience and the security fixes!

rocket-martue commented:
Hi @mayukojpn! Thanks for reviewing! 🙏 No worries about the timing at all!

You're absolutely right - I checked and `locales.php` is indeed a copy from GlotPress (currently based on commit 4e75ab6), while the GlotPress develop branch has newer updates (2880 lines vs our 2742 lines).

**However, I'd suggest keeping this PR focused on the CSRF security fix.**

Since the Wordfence vulnerability report is already public, I think it's better to:
1. ✅ Merge this CSRF fix first (security priority)
2. 📝 Create a separate issue to track the `locales.php` update

This way we can get the security fix out quickly without expanding the review scope.

**Regarding best practices for including external PHP files:**

Great question! The current code already has `if ( ! class_exists( 'GP_Locale' ) )` guards, which means if GlotPress is active, its classes will be used instead of the bundled copy. We could enhance this further:

1. **Current approach (bundled copy)** - Works standalone, needs periodic updates
2. **Conditional loading** - Explicitly check for GlotPress first
3. **Composer dependency** - Cleaner but adds complexity

I can create a follow-up issue to discuss updating `locales.php` and potentially improving the loading strategy. Would that work for you?

Thanks again for maintaining this plugin! 🚀

mayukojpn commented:
@rocket-martue Thank you very much for the comment!

To keep the review scope clear and manageable, I am thinking to apply the code copied from GlotPress alongside this patch. I am not entirely confident reviewing the security fix alone, 🥹 so this approach feels safer to me. Apologies if this slightly broadens the change, but I plan to address it shortly and keep it focused. 🙏 I don't think it takes too much extra time.

If you have any suggestions on comments or notes that should be added to PHP files copied from GlotPress, I would really appreciate your advice 🙏

mayukojpn committed:
Replace locales.php with latest GlotPress 4.0.3 upstream version

After review, decided to sync with the latest official GlotPress
version instead of modifying the outdated bundled copy.

- Replaced: Modified version based on commit 4e75ab6
- With: Unmodified copy from commit efe00cd (2025-06-25)
- Source: https://github.com/GlotPress/GlotPress/blob/efe00cd4834b85719378d663b3bfcabdb6259e7f/locales/locales.php

Related: Previous commit modified this file, but this replaces it entirely with the trusted upstream version.

mayukojpn reviewed (approved):
Looks good to me 🚀
